src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Models/Hl7v2TraceInfo.cs

```diff
@@ -82,16 +79,26 @@ public static Hl7v2TraceInfo CreateTraceInfo(Hl7v2Data hl7v2Data)
             return new Hl7v2TraceInfo(unusedSegments);
         }
 
-        private static int FindOffsetInSegment(string segmentValue, Hl7v2EncodingCharacters encodingCharacters, int fieldIndex, int componentIndex)
+        private static int FindOffsetInSegment(string segmentHeader, string segmentValue, Hl7v2EncodingCharacters encodingCharacters, int fieldIndex, int componentIndex)
         {
+            // All values($segmentValue and $fieldValue) need to be unescaped firstly (from "\\" back to "\"),
+            // or the length will be incorrectly calculated
+            segmentValue = SpecialCharProcessor.Unescape(segmentValue);
+
+            // MSH segment should be treated separately since the first '|' in MSH segment is a special field
+            if (segmentHeader == "MSH")
```

Comment:
string.equals?

Reply:
In this case `==` works equally as `Equals`, but I'll change it to `Equals` for consistency, thx :)

Reply:
Not sure is segment ID with lower case valid? If so, let's set stringcomparison.invariantcultureignorecase.